File: strategy/Strategy.ts

interface IOperatorStrategy {
    calculate: (a: number, b: number) => number;
}

class AddOperator implements IOperatorStrategy {
    public calculate(a: number, b: number): number {
        console.log("The result of " + a + " + " + b + " =");
        return a + b;
    }
}

class SubtractOperator implements IOperatorStrategy {
    public calculate(a: number, b: number): number {
        console.log("The result of " + a + " - " + b + " =");
        return a - b;
    }
}

class MultiplyOperator implements IOperatorStrategy {
    public calculate(a: number, b: number): number {
        console.log("The result of " + a + " * " + b + " =");
        return a * b;
    }
}

class DivideOperator implements IOperatorStrategy {
    public calculate(a: number, b: number): number {
        console.log("The result of " + a + " / " + b + " =");
        return a / b;
    }
}

class Calculator {
    private operator: IOperatorStrategy;
    constructor(operator: IOperatorStrategy) {
        this.operator = operator;
    }
    public setOperator(operator: IOperatorStrategy) {
        this.operator = operator;
    }

    public calculate(a: number, b: number): number {
        return this.operator.calculate(a, b);
    }
}

const testStrategy = () => {
    const calculator = new Calculator(new AddOperator());
    console.log(calculator.calculate(1, 2));
    calculator.setOperator(new SubtractOperator());
    console.log(calculator.calculate(1, 2));
    calculator.setOperator(new MultiplyOperator());
    console.log(calculator.calculate(1, 2));
    calculator.setOperator(new DivideOperator());
    console.log(calculator.calculate(1, 2));
}

testStrategy();

File: strategy/Strategy.py
from enum import Enum
# python 3.8 and above
from typing import Protocol, Dict

class IOperatorStrategy(Protocol):
    def calculate(self, a: int, b: int) -> int:
        pass

class AddStrategy(IOperatorStrategy):
    def calculate(self, a: int, b: int) -> int:
        print(f"The result of {a} + {b} = {a + b}")
        return a + b

class SubtractStrategy(IOperatorStrategy):
    def calculate(self, a: int, b: int) -> int:
        print(f"The result of {a} - {b} = {a - b}")
        return a - b

class MultiplyStrategy(IOperatorStrategy):
    def calculate(self, a: int, b: int) -> int:
        print(f"The result of {a} * {b} = {a * b}")
        return a * b
    
class DivideStrategy(IOperatorStrategy):
    def calculate(self, a: int, b: int) -> int:
        print(f"The result of {a} / {b} = {a / b}")
        return a / b
    
# a class calculator is our context
# set a private strategy field
class Calculator:
    def __init__(self, strategy: IOperatorStrategy):
        self._strategy = strategy
    def setStrategy(self, strategy: IOperatorStrategy):
        self._strategy = strategy
    def calculate(self, a: int, b: int) -> int:
        return self._strategy.calculate(a, b)

# test as a client 
if __name__ == "__main__":
    calculator = Calculator(AddStrategy())
    calculator.calculate(1, 2)
    
    calculator.setStrategy(SubtractStrategy())
    calculator.calculate(1, 2)
    
    calculator.setStrategy(MultiplyStrategy())
    calculator.calculate(1, 2)

    calculator.setStrategy(DivideStrategy())
    calculator.calculate(1, 2)